doc: Add security doc chapter (#7765)
nrfconnect/sdk-nrf: doc/nrf/security_chapter.rst

Security in NCS
###############

Introduction
************
The following chapters gives a brief introduction to core security features available on Nordic products.
The features are made available either as a built-in in modules, drivers and subsystems or as samples or application in nRF Connect SDK.

Secure boot
***********
NCS supports secure boot of application images enforced by the bootloader.
The secure boot utilizes signature validation and security hardware features to establish a root-of-trust during boot and to ensure the validity of the firmware that is booted.
There are two available bootloaders in NCS, the `nRF Secure Immutable Bootloader (NSIB)<http://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/samples/bootloader/README.html>` and a customized version of `MCUBoot <https://github.com/nrfconnect/sdk-mcuboot>`.
To enable NSIB the configuration :kconfig:option:`CONFIG_SECURE_BOOT` need to be enabled.
To enable MCUBoot the configuration :kconfig:option:`CONFING_MCUBOOT` need to be enabled.

> When enabling NSIB, it is possible to have an upgradable second-stage bootloader.

Reviewer: ```suggestion
When enabling NSIB, it is possible to have MCUboor serve as an upgradable second-stage bootloader.
```

Author: Reads better indeed, done!

Trusted Firmware-M
******************
The Trusted Firmware-M project (TF-M) is a reference design of the Arm Platform Security Arcitecture (PSA).

> NCS utilizes the security features of the Arm TrustZone techonolgy to configure secure peripherals and memory as well as to provide PSA functional APIs as secure services.

Reviewer: When reading it, it seems to me that is should rather be "TF-M utilizes the ..." or perhaps "Through TF-M, NCS utilizes the ..."? Or maybe I'm misunderstanding?

Author: Through TF-M, NCS utilizes..." seem more clear to me as well. Done

TF-M enables hardware supported separation of a Secure Processing Environment (SPE) and a Non-Secure Processing Environment (NSPE) that constitutes the Zephyr RTOS and the application.
To enable TF-M in a project the configuration :kconfig:option:`CONFIG_BUILD_WITH_TFM` need to be enabled.

TF-M adds a considerable code size increase when all its features are enabled.

> To tackle this the configuration :kconfig:option:`CONFIG_TFM_MINIMAL_CONFIG` was introduced.

Reviewer: ```suggestion
To tackle this the configuration :kconfig:option:`CONFIG_TFM_PROFILE_TYPE_MINIMAL` was introduced.
```

Author: Fixed

Enabling this configuration will enable APIs to generate random numbers in the non-secure image.

> TF-M using Arm TrustZone features is available on the following devices:

Reviewer: Does this make sense?
```suggestion
TF-M uses Arm TrustZone which is available on the following devices:
```

Author: Done

* nRF9160
* nRF5340

Hardware unique key
*******************

> The Nordic devices which feature the CryptoCell cryptographic accelerator allow the usage of a hardware unique key (HUK).

Reviewer: ```suggestion
The Nordic devices which feature the CryptoCell cryptographic accelerator allow the usage of a hardware unique key (HUK) for key derivation.
```

Author: Added this

The secure DFU mechanisms are handled by the MCUBoot bootloader, for more information please refer to the `MCUBoot documentation <http://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/mcuboot/index-ncs.html>`.

Crypto in NCS
*************
The cryptographic operations in NCS are handled by the :ref:`nrfxlib:nrf_security` module which is heavily configurable through Kconfig options.
The module can be enabled through the :kconfig:option:`CONFIG_NRF_SECURITY` Kconfig option.
The KConfig option :ref:`nrfxlib:nrf_security` allows the usage of `mbedTLS <https://tls.mbed.org/api/index.html>` and `PSA crypto <https://armmbed.github.io/mbed-crypto/html/index.html>` APIs for cryptographic operations and for random number generation in the application.
The :ref:`nrfxlib:nrf_security` acts as an orchestrator for the different cryptographic libraries available in the system.
These libraries include the binary versions of accelerated cryptographic libraries listed in :ref:`nrfxlib:crypto` and the open source implementation located in the `sdk-mbedtls <https://github.com/nrfconnect/sdk-mbedtls>`.

> The KConfig option :ref:`nrfxlib:nrf_security` prioritize the usage of the accelerated libraries by default when supported by the platform.

Reviewer: ```suggestion
The KConfig option :kconfig:option:`CONFIG_NRF_SECURITY` prioritize the usage of the accelerated libraries by default when supported by the platform.
```

?

For more information about the bootloaders, see :ref:`app_bootloaders`.

Trusted Firmware-M
******************

The Trusted Firmware-M project (TF-M) is a reference design of the Arm `Platform Security Architecture (PSA)`_.
Through TF-M, |NCS| utilizes the security features of the Arm TrustZone technology to configure secure peripherals and memory, and to provide PSA functional APIs as secure services.

TF-M enables hardware supported separation of a Secure Processing Environment (SPE) and a Non-Secure Processing Environment (NSPE) that constitutes the Zephyr RTOS, protocol stacks, and the application.
Enable TF-M in a project by enabling the :kconfig:option:`CONFIG_BUILD_WITH_TFM` option.

TF-M adds a considerable code size increase when all its features are enabled.
To minimize the code size increase, the configuration option :kconfig:option:`CONFIG_TFM_PROFILE_TYPE_MINIMAL` was introduced to provide the smallest possible build.
Enabling this configuration enables the APIs to generate random numbers in the non-secure image.

TF-M uses Arm TrustZone features available on the following devices:

* nRF9160
* nRF5340

> For more information about the TF-M, see the `TF-M documentation`_ and :ref:`ug_tfm`.

Author: The user guide has nearly identical information and we point it here as well so I will remove this from here.
